File: leeetcode/recursion/22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        
        output=[]
        
        def rec(left,right,stack,cand):
            if left==right==0:
                output.append(cand)
                return
            
            if left>0:
                rec(left-1,right,stack+1,cand+"(")
                
            if right>0 and stack> 0 :
                rec(left,right-1,stack-1,cand+")")  
                
                
        rec(n,n,0,"")
        
        return output

File: CPP/RECURSION/22. Generate Parentheses.cpp
class Solution {
public:
void paranth(int n,int left, int right,vector<string>&ans, string &temp)
{
    if (left+right==2*n)
    {
        ans.push_back(temp);
    }

    if(left<n)
    {
        temp.push_back('(');
        paranth(n,left+1,right,ans,temp);
        temp.pop_back();
    }

    if(right<left)
    {
        temp.push_back(')');
        paranth(n,left,right+1,ans,temp);
        temp.pop_back();
    }
}
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string temp;
        paranth(n,0,0,ans,temp);
        return ans;

        
    }
};
